torch_xla/csrc/xla_graph_executor.cpp

```diff
@@ -1136,7 +1142,9 @@ XLAGraphExecutor::CompilationResult XLAGraphExecutor::Compile(
     lowering_ctx.AddResult(root);
   }
   // Annotate HLO sharding selectively in the compuation.
-  bool is_sharded = ShardingUtil::SetHloSharding(&lowering_ctx);
+  // TODO(jonbolin): Always treat the computation as sharded when using SPMD.
+  bool is_sharded = ShardingUtil::SetHloSharding(&lowering_ctx) ||
+                    xla::sys_util::GetEnvBool("XLA_USE_SPMD", false);
```

Comment:
For our offline discussion, let's not rely on `XLA_USE_SPMD` -- it's meant to be replaced & for virtual device optimization. For implicit replication, we shouldn't rely on this flag -- instead let's see if we can have `InputHandler` decide.

Reply:
Following up on this - we aren't able to use InputHandler to track the handles across steps, since we only have access to the DataPtrs at that stage. We decided to rely on XLA_USE_SPMD in CreateTensorsData for now.